osquery/tables/system/windows/signature.cpp

```diff
@@ -0,0 +1,504 @@
+    break;
+  }
+
+  case TRUST_E_SUBJECT_NOT_TRUSTED:
+  default: {
+    result = SignatureInformation::Result::Untrusted;
+    break;
+  }
+  }
+
+  trust_provider_settings.dwStateAction = WTD_STATEACTION_CLOSE;
+  WinVerifyTrust(static_cast<HWND>(INVALID_HANDLE_VALUE),
+                 &authenticode_policy_provider,
+                 &trust_provider_settings);
+
+  if (verification_error) {
+    return Status(1, "Failed to verify the file signature");
+  }
+
+  return Status(0, "Ok");
+}
+
+Status getOriginalProgramName(SignatureInformation& signature_info,
+                              PCMSG_SIGNER_INFO signer_info) {
+  PCRYPT_ATTRIBUTE publisher_info_ptr = nullptr;
+
+  for (DWORD i = 0; i < signer_info->AuthAttrs.cAttr; i++) {
+    if (std::strcmp(SPC_SP_OPUS_INFO_OBJID,
+                    signer_info->AuthAttrs.rgAttr[i].pszObjId) == 0) {
+      publisher_info_ptr = &signer_info->AuthAttrs.rgAttr[i];
+      break;
+    }
+  }
+
+  if (publisher_info_ptr == nullptr) {
+    return Status(1, "The publisher information could not be found");
+  }
+
+  DWORD publisher_info_size;
+  if (!CryptDecodeObject(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                         SPC_SP_OPUS_INFO_OBJID,
+                         publisher_info_ptr->rgValue[0].pbData,
+                         publisher_info_ptr->rgValue[0].cbData,
+                         0,
+                         nullptr,
+                         &publisher_info_size)) {
+    return Status(1, "Failed to access the publisher information");
+  }
+
+  std::vector<std::uint8_t> publisher_info_blob_buffer;
+  publisher_info_blob_buffer.resize(static_cast<size_t>(publisher_info_size));
+  if (publisher_info_blob_buffer.size() !=
+      static_cast<size_t>(publisher_info_size)) {
+    return Status(1, "Memory allocation failure");
+  }
+
+  PSPC_SP_OPUS_INFO publisher_info_blob_ptr =
+      reinterpret_cast<PSPC_SP_OPUS_INFO>(publisher_info_blob_buffer.data());
+
+  if (!CryptDecodeObject(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                         SPC_SP_OPUS_INFO_OBJID,
+                         publisher_info_ptr->rgValue[0].pbData,
+                         publisher_info_ptr->rgValue[0].cbData,
+                         0,
+                         publisher_info_ptr,
+                         &publisher_info_size)) {
+    return Status(1, "Failed to decode the publisher information");
+  }
+
+  if (publisher_info_blob_ptr->pwszProgramName != nullptr) {
+    signature_info.original_program_name =
+        wstringToString(publisher_info_blob_ptr->pwszProgramName);
+  }
+
+  return Status(0, "Ok");
+}
+
+Status getCertificateInformation(SignatureInformation& signature_info,
+                                 PCCERT_CONTEXT certificate_context) {
+  PCRYPT_INTEGER_BLOB serial_number =
+      &certificate_context->pCertInfo->SerialNumber;
+
+  std::stringstream serial_number_string;
+  for (DWORD i = serial_number->cbData - 1; i < serial_number->cbData; i--) {
```

Comment:
Can you explain this loop?

Reply:
This is used to print the serial number backwards (as it is shown in the Authenticode dialog properties of Windows).

Since DWORD is unsigned, the loop uses the fact that according to the standard the value will wrap around once you decrement past zero.